src/main/java/build/buildfarm/common/config/Worker.java

```diff
@@ -63,4 +68,41 @@ public ExecutionPolicy[] getExecutionPolicies() {
       return new ExecutionPolicy[0];
     }
   }
+
+  public WorkerType getWorkerType() {
+    if (getCapabilities().isCas() && getCapabilities().isExecution()) {
+      return WorkerType.EXECUTE_AND_STORAGE;
+    }
+    if (getCapabilities().isCas()) {
+      return WorkerType.STORAGE;
+    }
+    return WorkerType.EXECUTE;
+  }
+
+  public Path getValidRoot() throws ConfigurationException {
+    addMissingRoot();
+    verifyRootConfiguration();
```

Comment:
I might misunderstand, but the order here seem to be wrong. 
`verifyRootConfiguration()` would verify if the configs for root is valid and `addMissingRoot` is creating the root if it's missing. It seems to me we want to do verify root option from config first first and then create

Reply:
The current verify method checks for both configuration error and missing dir on the filesystem which is why it's called after the adding of a directory. Perhaps it could be split into multiple verification methods.